pkg/p2p/server.go

```diff
@@ -0,0 +1,831 @@
+			acks = append(acks, &p2p.Ack{
+				Topic:   topic,
+				LastSeq: seq,
+			})
+			return true
+		})
+
+		if len(acks) == 0 {
+			continue
+		}
+
+		peer.metricsAckCount.Inc()
+		err := peer.sender.Send(ctx, p2p.SendMessageResponse{
+			Ack:        acks,
+			ExitReason: p2p.ExitReason_OK,
+		})
+		if err != nil {
+			log.Warn("sending response to peer failed", zap.Error(err))
+			if cerror.ErrPeerMessageInternalSenderClosed.Equal(err) {
+				peersToDeregister = append(peersToDeregister, peer)
+			}
+		}
+	}
+}
+
+func (m *MessageServer) deregisterPeer(ctx context.Context, peer *cdcPeer, err error) {
+	log.Info("Deregistering peer",
+		zap.String("sender", peer.PeerID),
+		zap.Int64("epoch", peer.Epoch),
+		zap.Error(err))
+
+	m.peerLock.Lock()
+	delete(m.peers, peer.PeerID)
+	m.peerLock.Unlock()
+	if err != nil {
+		peer.abortWithError(ctx, err)
+	}
+	m.acks.RemoveNode(peer.PeerID)
+}
+
+// We use an empty interface to hold the information on the type of the object
+// that we want to deserialize a message to.
+// We pass an object of the desired type, and use `reflect.TypeOf` to extract the type,
+// and then when we need it, we can use `reflect.New` to allocate a new object of this
+// type.
+type typeInformation = interface{}
+
+// SyncAddHandler registers a handler for messages in a given topic and waits for the operation
+// to complete.
+func (m *MessageServer) SyncAddHandler(
+	ctx context.Context,
+	topic string,
+	tpi typeInformation,
+	fn func(string, interface{}) error,
+) (<-chan error, error) {
+	doneCh, errCh, err := m.AddHandler(ctx, topic, tpi, fn)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	select {
+	case <-ctx.Done():
+		return nil, errors.Trace(ctx.Err())
+	case <-doneCh:
+	case <-m.closeCh:
+		return nil, cerror.ErrPeerMessageServerClosed.GenWithStackByArgs()
+	}
+	return errCh, nil
+}
+
+// AddHandler registers a handler for messages in a given topic.
+func (m *MessageServer) AddHandler(
+	ctx context.Context,
+	topic string,
+	tpi typeInformation,
+	fn func(string, interface{}) error) (chan struct{}, <-chan error, error) {
+	tp := reflect.TypeOf(tpi)
+
+	metricsServerRepeatedMessageCount := serverRepeatedMessageCount.MustCurryWith(prometheus.Labels{
+		"topic": topic,
+	})
+
+	poolHandle := m.pool.RegisterEvent(func(ctx context.Context, argsI interface{}) error {
+		args := argsI.(poolEventArgs)
+		sm := args.streamMeta
+		entry := args.entry
+		e := reflect.New(tp.Elem()).Interface()
+
+		lastAck := m.acks.Get(sm.SenderId, entry.GetTopic())
+		if lastAck >= entry.Sequence {
+			metricsServerRepeatedMessageCount.With(prometheus.Labels{
+				"from": sm.SenderAdvertisedAddr,
+			}).Inc()
+
+			log.Debug("skipping peer message",
+				zap.String("sender-id", sm.SenderId),
+				zap.String("topic", topic),
+				zap.Int64("skipped-Seq", entry.Sequence),
+				zap.Int64("last-ack", lastAck))
+			return nil
+		}
+
+		if lastAck != 0 && entry.Sequence > lastAck+1 {
```

Comment:
Is `entry.Sequence == lastAck+1` guaranteed by
- client receivers `acks` from server, and sends message with `seq = latest ack + 1`
- server maintains the correct ack corresponding to each message.

Reply:
`lastAck` here is actually the sequence number of the last message received, because we update the `ackManager` for each message. So in other words this is a requirement for the `seq` to be continuous, which can be guaranteed by the client locally.